File: src/parser_utils/output/__tests__/js_layout_generator/variable_declaration_precedence_test.ml
(*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

open Ast_builder
open Layout_test_utils
module S = Ast_builder.Statements
module E = Ast_builder.Expressions
module L = Layout_builder

let test ctxt =
  let seq = E.sequence [E.identifier "y"; E.identifier "z"] in
  let ast = S.variable_declaration [S.variable_declarator "x" ~init:seq] in
  assert_layout_of_statement
    ~ctxt
    L.(
      loc
        (loc
           (fused
              [
                atom "var";
                atom " ";
                loc
                  (fused
                     [
                       loc (id "x");
                       pretty_space;
                       atom "=";
                       pretty_space;
                       wrap_in_parens (expression seq);
                     ]
                  );
                atom ";";
              ]
           )
        )
    )
    ast;

  let ast =
    let init = E.assignment (Patterns.identifier "y") (E.identifier "z") in
    S.variable_declaration [S.variable_declarator "x" ~init]
  in
  assert_layout_of_statement
    ~ctxt
    L.(
      loc
        (loc
           (fused
              [
                atom "var";
                atom " ";
                loc
                  (fused
                     [
                       loc (id "x");
                       pretty_space;
                       atom "=";
                       pretty_space;
                       loc (fused [loc (id "y"); pretty_space; atom "="; pretty_space; loc (id "z")]);
                     ]
                  );
                atom ";";
              ]
           )
        )
    )
    ast;

  let fn_ast = E.function_ () in
  let ast = S.variable_declaration [S.variable_declarator "x" ~init:fn_ast] in
  assert_layout_of_statement
    ~ctxt
    L.(
      loc
        (loc
           (fused
              [
                atom "var";
                atom " ";
                loc (fused [loc (id "x"); pretty_space; atom "="; pretty_space; expression fn_ast]);
                atom ";";
              ]
           )
        )
    )
    ast

File: rust_port/crates/flow_parser_utils_output/src/js_layout_generator_test/variable_declaration_precedence_test.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use flow_parser_utils::ast_builder::expressions as E;
use flow_parser_utils::ast_builder::statements as S;
use flow_parser_utils::ast_builder::*;

use crate::layout_test_utils::layout_builder as L;
use crate::layout_test_utils::*;

#[test]
fn test() {
    let seq = E::sequence(
        None,
        None,
        vec![
            E::identifier(None, None, "y"),
            E::identifier(None, None, "z"),
        ],
    );
    let ast = S::variable_declaration(
        None,
        None,
        None,
        vec![S::variable_declarator(None, Some(seq.clone()), None, "x")],
    );
    assert_layout_of_statement(
        None,
        L::loc(
            None,
            L::loc(
                None,
                L::fused(vec![
                    L::atom("var"),
                    L::atom(" "),
                    L::loc(
                        None,
                        L::fused(vec![
                            L::loc(None, L::id(None, "x")),
                            L::pretty_space(),
                            L::atom("="),
                            L::pretty_space(),
                            L::wrap_in_parens(L::expression(None, None, &seq)),
                        ]),
                    ),
                    L::atom(";"),
                ]),
            ),
        ),
        &ast,
    );

    let ast = {
        let init = E::assignment(
            None,
            None,
            patterns::identifier(None, None, "y"),
            None,
            E::identifier(None, None, "z"),
        );
        S::variable_declaration(
            None,
            None,
            None,
            vec![S::variable_declarator(None, Some(init), None, "x")],
        )
    };
    assert_layout_of_statement(
        None,
        L::loc(
            None,
            L::loc(
                None,
                L::fused(vec![
                    L::atom("var"),
                    L::atom(" "),
                    L::loc(
                        None,
                        L::fused(vec![
                            L::loc(None, L::id(None, "x")),
                            L::pretty_space(),
                            L::atom("="),
                            L::pretty_space(),
                            L::loc(
                                None,
                                L::fused(vec![
                                    L::loc(None, L::id(None, "y")),
                                    L::pretty_space(),
                                    L::atom("="),
                                    L::pretty_space(),
                                    L::loc(None, L::id(None, "z")),
                                ]),
                            ),
                        ]),
                    ),
                    L::atom(";"),
                ]),
            ),
        ),
        &ast,
    );

    let fn_ast = E::function_(None, None, None, None, None, None);
    let ast = S::variable_declaration(
        None,
        None,
        None,
        vec![S::variable_declarator(
            None,
            Some(fn_ast.clone()),
            None,
            "x",
        )],
    );
    assert_layout_of_statement(
        None,
        L::loc(
            None,
            L::loc(
                None,
                L::fused(vec![
                    L::atom("var"),
                    L::atom(" "),
                    L::loc(
                        None,
                        L::fused(vec![
                            L::loc(None, L::id(None, "x")),
                            L::pretty_space(),
                            L::atom("="),
                            L::pretty_space(),
                            L::expression(None, None, &fn_ast),
                        ]),
                    ),
                    L::atom(";"),
                ]),
            ),
        ),
        &ast,
    );
}
